device/tt_silicon_driver.cpp

```diff
@@ -2140,23 +2176,27 @@ void tt_SiliconDevice::write_device_memory(const void *mem_ptr, uint32_t size_in
     //     target.chip, target.x, target.y, address, size_in_bytes, small_access);
 
     std::int32_t tlb_index = 0;
-    std::optional<std::tuple<std::uint32_t, std::uint32_t>> tlb_data = std::nullopt;
+    std::optional<std::tuple<std::uint64_t, std::uint64_t>> tlb_data = std::nullopt;
     if(tlbs_init) {
         tlb_index = map_core_to_tlb(tt_xy_pair(target.x, target.y));
         tlb_data = dev->get_architecture_implementation()->describe_tlb(tlb_index);
     }
 
     if (tlb_data.has_value() && address_in_tlb_space(address, size_in_bytes, tlb_index, std::get<1>(tlb_data.value()), target.chip)) {
         auto [tlb_offset, tlb_size] = tlb_data.value();
-        write_block(dev, tlb_offset + address % tlb_size, size_in_bytes, buffer_addr, m_dma_buf_size);
+        if (dev->bar4_mapping != nullptr && tlb_size > BAR0_BH_SIZE) {
```

Comment:
There are 1 MiB, 2 MiB, 16 MiB, and 4 GiB TLB sizes supported by UMD.  What is the rationale behind comparing TLB size to BAR0 size?

Reply:
I see what it is doing (add this constant to the address, which read/write block treats as a flag, of sorts).  This has the potential for confusion for two reasons:
* WH also has BAR4, but a different name (`system_reg_mapping`).  `dev->bar4_mapping` is BH specific, but this may not be obvious to someone reading the code.
* TLB size and BAR size are different things.

It makes sense from the standpoint that writes "beyond" BAR0 go to BAR4.

Not suggesting that you change anything here, but for your consideration:

 Would `if (dev->bar4_mapping != nullptr && tlb_size == blackhole::TLB_4G_SIZE) {` be clearer?  Is there opportunity for different/improved `write_block` and `read_block` functions?

Reply:
I completely agree that comparison between TLB size and BAR0 size was not clear, problem in tt_silicon_driver in general is that arch specific stuff is not accessible in that file. I have defined TLB_4G_SIZE and changed the code how you suggested. Regarding write/read block, it could be possibly moved to arch specific files and then just exposed to tt silicon driver, because with different configs we want to choose TLB address in different way, as seen above